### DISCUSSION TITLE
USing DAFoam for co jet flow - airfoil coordinates modfication req

### DISCUSSION THREAD

**Discussion by zonexo** [mdolab/dafoam#694] [General]
Hi, I'm thinking of using DAFoam for my optimization problem but I'm not sure if it would work. I am trying to optimize co-jet flow [airfoil.](http://coflowjet.com/). So the thing is that the airfoil geometry is slight modified with part of the top suction side being lower and there's coming out and going into the slots. In essense, I need to:

1. modify the baseline airfoil shape - standard DAFoam
2. use a python or some codes to modify the suction side coordinates of the new airfoil - I have the formula - non- standard
3. run DAFoam, get cl, cd etc and go back to step 1

Btw, in my problem, I also need to impose constant mass rate or velocity or total/static pressure at the injection and suction slot.

So can I use DAFoam? Key thing is tt I need to modify the airfoil coordinates again after DAFoam has modified the airfoil. Another thing is the BC.

Thanks!

**Comment by friedenhe** [comment]
I am not sure I quite understand your setup. Can you be more specific, or perhaps have a diagram to better explain this new concept?

**Reply by zonexo** [reply]
![co flow jet](https://github.com/user-attachments/assets/4dfa8a7b-26fc-4548-9c7b-5cbd163bad8f)

Sure. So I've attached the co-flow jet design, vs org airfoil. It has 2 slots for air in and output. Part of the top part of the org airfoil has to be shifted down slightly, as shown.

From my understandiing, not sure if it's correct, that DAFoam optimizes by changing the airfoil shape and the grid deforms accordng to the shape change. It does this until some constraint is satisfied.

So for my case, after the airfoil shape is changed, I need to use a code to modify the airfoil coordinates to shift down the top part before DAFoam does its calculation. So is this possible? I prefer to use a Python code to modify if possible, Of cos, else a C code is also ok.

Thanks!

**Comment by friedenhe** [comment]
I am still a bit confused about your setup. What are your objective function, design variables, and constraints? What is the relationship between the top and bottom of the picture you posted? Are you trying to change the shape for the top or the bottom airfoil? Are you trying to reduce the drag for the top airfoil, or maximize the lift for the bottom one? Are these two airfoils represent different working conditions?

**Comment by zonexo** [comment]
Hi, sorry for the confusion. The top is the standard airfoil, while the bottom is a co-flow jet airfoil. my focus is on the co-flow jet. Co-flow jet has part of its top surface shifted down and there's 2 slots with fixed mass flow jet as boundary condition.

I've different obj but let's say I'm trying to maximize the cl/cd ratio of a co-flow jet, as shown at the bottom pic, by changing its shape. So maybe best shape for max cl/cd in 10,000 iterations. The operating conditions like inflow vel etc are fixed, except for the airfoil shape, which changes as DAFoam optimizes I supposed. 

I don't think DAFoam can optimize this co-flow jet jet directly. So is it possible to let DAFoam change the shape of the org airfoil, then I insert a code to shift the top surface down to convert it to a co-flow jet, then DAFoam or OpenFOAM take over again to deform the grid and run the simulation. This process is then repeated.

Hope it's clearer now. Thanks for the help.

**Comment by friedenhe** [comment]
Based on your description, I believe DAFoam can directly optimize the co-flow jet airfoil. There is no need to first optimize the original airfoil and then manually shift down the top surface. Let me know if you have any concerns about directly optimizing the bottom airfoil.

**Comment by zonexo** [comment]
Oh that's great! I have attached another diagram.

![co flow jet DAFoam](https://github.com/user-attachments/assets/f2f5a9c4-3979-49b4-8740-539fe67d9978)

So to be more precise, is it possible to change and optimize only the shape along the blue line, while keeping red slot region length fixed and perpendicular to its adjacent region, as shown? 

If so, what tutorial should I follow to achieve this?

Thanks

**Comment by friedenhe** [comment]
Yes, you can directly optimize the co-jet airfoil. Then, you need to impose two constraints: the first is to enforce the length of the slot and the other is to ensure orthogonality. Unfortunately, these two constraints are not in pyGeo yet. We can implement it, but it may take a while due to the limited manpower.

**Reply by zonexo** [reply]
Sure, thanks for the reply. I'll play around 1st with the code to see what I can do with it. 

Btw, last time I asked the qn if I can run a python to shift part of the airfoil down myself right after DAFoam produces a new airfoil shape. Can it be done? Then, I can implement to change the airfoil to a coflow jet airfoil based on the 2 constraints, then DAFoam can take over and run OF after that.

**Reply by friedenhe** [reply]
> Btw, last time I asked the qn if I can run a python to shift part of the airfoil down myself right after DAFoam produces a new airfoil shape. Can it be done? 

I am not sure how this will work because once you manually shift down the airfoil after DAFoam produces a new airfoil shape, it essentially changes the volume mesh. And you need to somehow let DAFoam know this change, and more importantly, the gradients need to be passed through this manual change. This is not an easy task.
